Update base client to v0.7.0

camille-bouvy-frequenz opened the pull request:
Update base client from 0.6.1 to v0.7.0, and upgrade the code accordingly.

daniel-zullo-frequenz committed:
Update examples to use asyncio event loop

The user had tried to instantiate the client outside the
asyncio event loop, but it must be created within the loop
to handle asynchronous calls correctly.

This change updates the examples to use async functions
and runs them within an asyncio event loop.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

Commit:
Update base client version

Signed-off-by: camille-bouvy-frequenz <camille.bouvy@frequenz.com>

shsms commented on src/frequenz/client/electricity_trading/_client.py:
```diff
@@ -13,10 +13,12 @@
 # pylint: disable=no-member
 from frequenz.api.electricity_trading.v1 import electricity_trading_pb2
 from frequenz.api.electricity_trading.v1.electricity_trading_pb2_grpc import (
+    ElectricityTradingServiceAsyncStub,
```
@llucax, we can't import the async stub like this, right?  Should this only be under `if typing.TYPE_CHECKING:`?

camille-bouvy-frequenz replied:
This was done [here](https://github.com/frequenz-floss/frequenz-client-base-python/releases/tag/v0.7.0) in the example section of the upgrading section of the new base client release. But since it's just an example I don't know if it was tested

shsms replied:
I just tried as well, can't find an easy solution.  I guess we need to wait for @llucax

Marenz replied:
try just importing 
```
from frequenz.api.electricity_trading.v1 import electricity_trading_pb2_grpc
```

and then just using `electricity_trading_pb2_grpc.ElectricityTradingServiceAsyncStub` with a global setting to disable the no-member check in pyproject.toml
```
[tool.pylint.messages_control]
disable = [
  [..]
  # Checked by mypy
  "no-member",
  [..]
]
```

llucax replied:
Thanks @Marenz for the summary, that should work.

I just updated the release notes (once again) to hopefully provide instructions that finally work.

Marenz replied:
Correction: I wrote `electricity_trading_pb2` but meant `electricity_trading_pb2_grpc`, it is updated in my comment

camille-bouvy-frequenz replied:
I still get an error from `pytest_min`: 
```
ERROR AttributeError: module 'frequenz.api.electricity_trading.v1.electricity_trading_pb2_grpc' has no attribute 'ElectricityTradingServiceAsyncStub'
```

shsms commented on src/frequenz/client/electricity_trading/_client.py:
```diff
@@ -123,7 +125,11 @@ def __init__(
         if not hasattr(
             self, "_initialized"
         ):  # Prevent re-initialization of existing instances
-            super().__init__(server_url, ElectricityTradingServiceStub, connect=connect)
+            super().__init__(server_url, connect=connect)
+            self._stub = cast(
+                ElectricityTradingServiceAsyncStub,
+                ElectricityTradingServiceStub(self.channel),
+            )
```
It seems there's another problem here.

We have to create the `_stub` when we are connecting.  So if `connect=False`, we can't create the stub immediately, because there's no channel.

And we're going to have to override `connect`, to call super.connect(), then create the stub.

@llucax this is also not covered in the upgrade notes.

I'll try to make a working patch for this now.

shsms replied:
The new base client's interface feels inconsistent compared to the older one because we have to override connect, but with that, it works for now.

shsms committed:
Rename variable to from `_` to `client`

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Replace `dict[str, Any]` with `SetupParams` dataclass in tests

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Create a stub only when `connect=True`

Also import the `ElectricityTradingServiceAsyncStub` type only when
`TYPE_CHECKING`.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Check if stub is available before trying to fetch it

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Override the `BaseApiClient.connect` method to also create a stub

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

Commit:
Update Release Notes and clear leftover features from release v0.2.5

Signed-off-by: camille-bouvy-frequenz <camille.bouvy@frequenz.com>
